File: internal/insights/tool_analytics/aggregator.go
package tool_analytics

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"time"

	"claude-devtools/internal/discovery"
)

// medianU64 returns the median of a sorted-in-place slice. Empty → 0.
// Mirrors aggregator.rs::median_u64.
func medianU64(samples []uint64) uint64 {
	if len(samples) == 0 {
		return 0
	}
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	mid := len(samples) / 2
	if len(samples)%2 == 0 {
		return (samples[mid-1] + samples[mid]) / 2
	}
	return samples[mid]
}

// finalize converts the raw stats map into sorted ToolUsageSummary slice.
// Sorted by call_count descending. Mirrors aggregator.rs::finalize.
func finalize(stats map[string]*toolStats) []ToolUsageSummary {
	out := make([]ToolUsageSummary, 0, len(stats))
	for name, s := range stats {
		avgDuration := 0.0
		if len(s.durationSamples) > 0 {
			sum := 0.0
			for _, d := range s.durationSamples {
				sum += d
			}
			avgDuration = sum / float64(len(s.durationSamples))
		}
		medianTokens := medianU64(s.tokenSamples)
		successRate, errorRate := 0.0, 0.0
		if s.callCount > 0 {
			successRate = float64(s.successCount) / float64(s.callCount)
			errorRate = float64(s.errorCount) / float64(s.callCount)
		}
		out = append(out, ToolUsageSummary{
			ToolName:        name,
			CallCount:       s.callCount,
			SuccessCount:    s.successCount,
			ErrorCount:      s.errorCount,
			SuccessRate:     successRate,
			ErrorRate:       errorRate,
			AvgDurationMs:   avgDuration,
			MedianTokenCost: medianTokens,
		})
	}
	sort.Slice(out, func(i, j int) bool {
		return out[i].CallCount > out[j].CallCount
	})
	return out
}

// resolveProjectDir resolves the project directory path from a project ID.
// Mirrors aggregator.rs::resolve_project_dir.
func resolveProjectDir(projectID string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("cannot resolve home directory")
	}
	claudeDir := home + "/.claude"
	projectsDir := discovery.GetProjectsBasePath(claudeDir)

	baseID := projectID
	if i := strings.Index(projectID, "::"); i >= 0 {
		baseID = projectID[:i]
	}
	dir := projectsDir + "/" + baseID
	info, err := os.Stat(dir)
	if err != nil || !info.IsDir() {
		return "", fmt.Errorf("project directory not found: %s", baseID)
	}
	return dir, nil
}

// ComputeToolAnalytics scans sessions in a project and returns per-tool usage
// stats. days is clamped to [1,90]. Mirrors aggregator.rs::compute_tool_analytics.
func ComputeToolAnalytics(projectID string, days uint32) (*ToolAnalyticsResponse, error) {
	projectDir, err := resolveProjectDir(projectID)
	if err != nil {
		return nil, err
	}
	if days < 1 {
		days = 1
	}
	if days > 90 {
		days = 90
	}

	nowMs := float64(time.Now().UnixMilli())
	cutoffMs := nowMs - float64(days)*86_400_000.0

	entries, err := os.ReadDir(projectDir)
	if err != nil {
		return nil, fmt.Errorf("read dir: %w", err)
	}

	stats := make(map[string]*toolStats)
	var scannedSessions uint32

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".jsonl") {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		modifiedMs := float64(info.ModTime().UnixMilli())
		if modifiedMs < cutoffMs {
			continue
		}
		scannedSessions++
		scanSession(projectDir+"/"+entry.Name(), stats)
	}

	tools := finalize(stats)
	var totalCalls, totalErrors uint32
	for _, t := range tools {
		totalCalls += t.CallCount
		totalErrors += t.ErrorCount
	}

	return &ToolAnalyticsResponse{
		Tools:           tools,
		TotalCalls:      totalCalls,
		TotalErrors:     totalErrors,
		ScannedSessions: scannedSessions,
	}, nil
}

// ComputeToolTimeHeatmap returns a 7×24 heatmap of tool usage bucketed by
// local (weekday, hour). Mirrors aggregator.rs::compute_tool_time_heatmap.
func ComputeToolTimeHeatmap(projectID string, days uint32, toolFilter string) (*ToolTimeHeatmapResponse, error) {
	projectDir, err := resolveProjectDir(projectID)
	if err != nil {
		return nil, err
	}
	if days < 1 {
		days = 1
	}
	if days > 90 {
		days = 90
	}

	nowMs := float64(time.Now().UnixMilli())
	cutoffMs := nowMs - float64(days)*86_400_000.0

	entries, err := os.ReadDir(projectDir)
	if err != nil {
		return nil, fmt.Errorf("read dir: %w", err)
	}

	buckets := make(map[heatmapKey]*heatmapCellAcc)
	// BTreeSet equivalent: use a map then sort keys.
	allToolsSet := make(map[string]struct{})

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".jsonl") {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		modifiedMs := float64(info.ModTime().UnixMilli())
		if modifiedMs < cutoffMs {
			continue
		}
		fullPath := projectDir + "/" + entry.Name()

		// Collect all tool names regardless of filter.
		unfiltered := make(map[heatmapKey]*heatmapCellAcc)
		scanSessionHeatmap(fullPath, unfiltered, "")
		for _, cell := range unfiltered {
			for name := range cell.perTool {
				allToolsSet[name] = struct{}{}
			}
		}

		// Real scan with filter for bucket totals.
		scanSessionHeatmap(fullPath, buckets, toolFilter)
	}

	// Build ordered 7×24 cells (day 0..6, hour 0..23).
	cells := make([]ToolTimeHeatmapCell, 0, 7*24)
	for day := uint8(0); day < 7; day++ {
		for hour := uint8(0); hour < 24; hour++ {
			key := makeHeatmapKey(day, hour)
			cell := buckets[key]

			var callCount uint32
			var topTool *string
			if cell != nil {
				callCount = cell.total
				// Find top tool deterministically: max count, then alpha on ties.
				var bestName string
				var bestCount uint32
				for name, cnt := range cell.perTool {
					if cnt > bestCount || (cnt == bestCount && name < bestName) {
						bestName = name
						bestCount = cnt
					}
				}
				if bestName != "" {
					s := bestName
					topTool = &s
				}
			}
			cells = append(cells, ToolTimeHeatmapCell{
				DayOfWeek: day,
				Hour:      hour,
				CallCount: callCount,
				TopTool:   topTool,
			})
		}
	}

	var totalCalls uint32
	for _, c := range cells {
		totalCalls += c.CallCount
	}

	// Sort tool names (BTreeSet iteration order = sorted).
	allToolNames := make([]string, 0, len(allToolsSet))
	for name := range allToolsSet {
		allToolNames = append(allToolNames, name)
	}
	sort.Strings(allToolNames)

	return &ToolTimeHeatmapResponse{
		Cells:      cells,
		TotalCalls: totalCalls,
		ToolNames:  allToolNames,
	}, nil
}

File: src-tauri/src/insights/tool_analytics/aggregator.rs
use std::collections::HashMap;
use std::path::PathBuf;

use crate::config::root;
use crate::discovery::path_decoder;

use super::scanner::{scan_session, scan_session_heatmap};
use super::shared::{
    heatmap_key, HeatmapCellAcc, HeatmapKey, ToolAnalyticsResponse, ToolStats, ToolTimeHeatmapCell,
    ToolTimeHeatmapResponse, ToolUsageSummary,
};

pub(super) fn median_u64(samples: &mut [u64]) -> u64 {
    if samples.is_empty() {
        return 0;
    }
    samples.sort_unstable();
    let mid = samples.len() / 2;
    if samples.len() % 2 == 0 {
        (samples[mid - 1] + samples[mid]) / 2
    } else {
        samples[mid]
    }
}

pub(super) fn finalize(stats: HashMap<String, ToolStats>) -> Vec<ToolUsageSummary> {
    let mut out: Vec<ToolUsageSummary> = stats
        .into_iter()
        .map(|(name, mut s)| {
            let call_count = s.call_count;
            let avg_duration = if s.duration_samples.is_empty() {
                0.0
            } else {
                s.duration_samples.iter().sum::<f64>() / s.duration_samples.len() as f64
            };
            let median_tokens = median_u64(&mut s.token_samples);
            let (success_rate, error_rate) = if call_count == 0 {
                (0.0, 0.0)
            } else {
                (
                    s.success_count as f64 / call_count as f64,
                    s.error_count as f64 / call_count as f64,
                )
            };
            ToolUsageSummary {
                tool_name: name,
                call_count,
                success_count: s.success_count,
                error_count: s.error_count,
                success_rate,
                error_rate,
                avg_duration_ms: avg_duration,
                median_token_cost: median_tokens,
            }
        })
        .collect();
    out.sort_by(|a, b| b.call_count.cmp(&a.call_count));
    out
}

fn resolve_project_dir(project_id: &str) -> Result<PathBuf, String> {
    let claude_dir = root::claude_dir()?;
    let projects_dir = path_decoder::get_projects_base_path(&claude_dir);
    let base_id = match project_id.find("::") {
        Some(idx) => &project_id[..idx],
        None => project_id,
    };
    let project_dir = projects_dir.join(base_id);
    if !project_dir.is_dir() {
        return Err(format!("Project directory not found: {base_id}"));
    }
    Ok(project_dir)
}

pub fn compute_tool_analytics(
    project_id: &str,
    days: u32,
) -> Result<ToolAnalyticsResponse, String> {
    let project_dir = resolve_project_dir(project_id)?;
    let days = days.clamp(1, 90);

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| e.to_string())?
        .as_millis() as f64;
    let cutoff_ms = now_ms - (days as f64) * 86_400_000.0;

    // Go's os.ReadDir returns entries sorted by filename; mirror that.
    let mut entries: Vec<_> = std::fs::read_dir(&project_dir)
        .map_err(|e| e.to_string())?
        .flatten()
        .collect();
    entries.sort_by_key(|e| e.file_name());

    let mut stats: HashMap<String, ToolStats> = HashMap::new();
    let mut scanned_sessions: u32 = 0;

    for entry in entries {
        let fname = entry.file_name();
        let fname = fname.to_string_lossy();
        if !fname.ends_with(".jsonl") {
            continue;
        }
        let modified_ms = entry
            .metadata()
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_millis() as f64)
            .unwrap_or(0.0);
        if modified_ms < cutoff_ms {
            continue;
        }
        scanned_sessions += 1;
        let _ = scan_session(&entry.path(), &mut stats);
    }

    let tools = finalize(stats);
    let total_calls: u32 = tools.iter().map(|t| t.call_count).sum();
    let total_errors: u32 = tools.iter().map(|t| t.error_count).sum();

    Ok(ToolAnalyticsResponse {
        tools,
        total_calls,
        total_errors,
        scanned_sessions,
    })
}

pub fn compute_tool_time_heatmap(
    project_id: &str,
    days: u32,
    tool_filter: Option<&str>,
) -> Result<ToolTimeHeatmapResponse, String> {
    let project_dir = resolve_project_dir(project_id)?;
    let days = days.clamp(1, 90);

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| e.to_string())?
        .as_millis() as f64;
    let cutoff_ms = now_ms - (days as f64) * 86_400_000.0;

    // Go's os.ReadDir returns entries sorted by filename; mirror that.
    let mut entries: Vec<_> = std::fs::read_dir(&project_dir)
        .map_err(|e| e.to_string())?
        .flatten()
        .collect();
    entries.sort_by_key(|e| e.file_name());

    let mut buckets: HashMap<HeatmapKey, HeatmapCellAcc> = HashMap::new();
    // Also collect the full tool name set across the range (ignoring filter)
    // so the dropdown always shows every tool, even if filter removes rows.
    let mut all_tools: std::collections::BTreeSet<String> = std::collections::BTreeSet::new();

    for entry in entries {
        let fname = entry.file_name();
        let fname = fname.to_string_lossy();
        if !fname.ends_with(".jsonl") {
            continue;
        }
        let modified_ms = entry
            .metadata()
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_millis() as f64)
            .unwrap_or(0.0);
        if modified_ms < cutoff_ms {
            continue;
        }
        let path = entry.path();

        // Collect tool names regardless of filter.
        let mut unfiltered: HashMap<HeatmapKey, HeatmapCellAcc> = HashMap::new();
        let _ = scan_session_heatmap(&path, &mut unfiltered, None);
        for cell in unfiltered.values() {
            for name in cell.per_tool.keys() {
                all_tools.insert(name.clone());
            }
        }

        // Real scan with filter for bucket totals.
        let _ = scan_session_heatmap(&path, &mut buckets, tool_filter);
    }

    let mut cells: Vec<ToolTimeHeatmapCell> = Vec::with_capacity(7 * 24);
    for day in 0u8..7 {
        for hour in 0u8..24 {
            let cell = buckets.remove(&heatmap_key(day, hour)).unwrap_or_default();
            // Deterministic top tool: max count, tie-break on smallest name
            // (matches Go: `cnt > best || (cnt == best && name < bestName)`).
            let mut best: Option<(&str, u32)> = None;
            for (name, &cnt) in cell.per_tool.iter() {
                match best {
                    Some((bn, bc)) if bc > cnt || (bc == cnt && bn <= name.as_str()) => {}
                    _ => best = Some((name.as_str(), cnt)),
                }
            }
            let top_tool = best.map(|(name, _)| name.to_string());
            cells.push(ToolTimeHeatmapCell {
                day_of_week: day,
                hour,
                call_count: cell.total,
                top_tool,
            });
        }
    }

    let total_calls: u32 = cells.iter().map(|c| c.call_count).sum();

    Ok(ToolTimeHeatmapResponse {
        cells,
        total_calls,
        tool_names: all_tools.into_iter().collect(),
    })
}
